segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+            if (isMainSegmentDeleted()) {
+                // Verify if the main segment has been deleted and stop now if needed.
+                return handleMainSegmentDeleted(timeout);
+            }
+
+            long lastSnapshotOffset = getLastSnapshotOffset();
+            int readLength = (int) Math.min(Integer.MAX_VALUE, this.attributeSegment.get().getLength() - lastSnapshotOffset);
+            CompletableFuture<AttributeCollection> r = new CompletableFuture<>();
+            if (readLength == 0) {
+                // Nothing to read.
+                r.complete(new AttributeCollection());
+            } else {
+                AsyncReadResultProcessor.process(
+                        StreamSegmentStorageReader.read(this.attributeSegment.get().handle, lastSnapshotOffset, readLength, this.config.getReadBlockSize(), this.storage),
+                        new AttributeSegmentReader(r, timeout),
+                        this.executor);
+            }
+            return r;
+        }, this.executor);
+        return Futures.exceptionallyCompose(result,
+                ex -> {
+                    ex = Exceptions.unwrap(ex);
+                    if (ex instanceof IOException || ex instanceof StreamSegmentTruncatedException) {
+                        ex = new DataCorruptionException(
+                                String.format("Unable to parse AttributeSegment. LastSnapshot = (Offset=%d, Length=%d), Known Segment Length = %d.",
+                                        getLastSnapshotOffset(), getLastSnapshotLength(), this.attributeSegment.get().getLength()), ex);
+                    }
+                    return Futures.failedFuture(ex);
+                });
+    }
+
+    /**
+     * Creates a new Snapshot, writes it to the AttributeSegment and updates the main Segment's attributes with its location.
+     *
+     * @param newAttributes Any new Attributes to include in this Snapshot that would not already be part of it.
+     * @param mustComplete  Whether the update of the main Segment's attributes must execute successfully in order for this
+     *                      operation to be considered complete. For auto-generated Snapshots, since the Snapshot itself is
+     *                      already written to the AttributeSegment, it is not mandatory for the OperationLog Add &
+     *                      Attribute Segment Truncation operations to complete, as there will be no data loss. However, for
+     *                      Snapshots generated as part of Seals, all of those must complete in order for the operation to
+     *                      be considered successful.
+     * @param timeout       Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the operation has completed successfully.
+     */
+    private CompletableFuture<Void> createSnapshot(AttributeCollection newAttributes, boolean mustComplete, Duration timeout) {
```

Comment:
Can we add an info message to indicate that we are creating a snapshot? Or is it going to be too frequent?

Reply:
There is a debug message at line 346. I don't want to make this info since this is per segment, so we might end up polluting the logs with this.